From e6e69a3a417baa4dc76a186b272059fe4e7fe360 Mon Sep 17 00:00:00 2001
From: Avery Pennarun <apenwarr@gmail.com>
Date: Wed, 24 Sep 2014 23:05:39 -0400
Subject: [PATCH] hostapd: exit with error code if underlying interface
 disappears.

hostapd tries to recover when this happens, but it doesn't work and it
leaves the interface in a nonfunctional state.  Instead, just abort and let
the babysitter recover.
---
 src/ap/drv_callbacks.c       | 7 +++++++
 src/drivers/driver_nl80211.c | 2 +-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index f86574f..ee26903 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -1135,6 +1135,13 @@ void wpa_supplicant_event(void *ctx, enum wpa_event_type event,
 		hostapd_channel_list_updated(
 			hapd->iface, data->channel_list_changed.initiator);
 		break;
+	case EVENT_INTERFACE_STATUS:
+		if (data->interface_status.ievent == EVENT_INTERFACE_REMOVED) {
+			wpa_printf(MSG_ERROR, "Interface removed by someone; exiting.\n");
+			hapd->iface->interfaces->terminate_on_error++;
+			eloop_terminate();
+		}
+		break;
 #endif /* NEED_AP_MLME */
 	default:
 		wpa_printf(MSG_DEBUG, "Unknown event %d", event);
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index c114e3b..8f2ded6 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -1253,7 +1253,7 @@ static void wpa_driver_nl80211_event_rtm_newlink(void *ctx,
 				   "event since interface %s is up", namebuf);
 			return;
 		}
-		wpa_printf(MSG_DEBUG, "nl80211: Interface down");
+		wpa_printf(MSG_WARNING, "nl80211: Interface down");
 		if (drv->ignore_if_down_event) {
 			wpa_printf(MSG_DEBUG, "nl80211: Ignore interface down "
 				   "event generated by mode change");
-- 
2.1.0.rc2.206.gedb03e5

